### PR TITLE
feat: push benefit method

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/api/entities/DividendDistribution/index.ts [hunk 1]
```diff
@@ -95,13 +97,23 @@ export class DividendDistribution extends CorporateAction {
...
 
   /**
    * Claim the dividends corresponding to the current Identity
    */
   public claim: ProcedureMethod<void, void>;
 
+  /**
+   * Push the capital distribution benefit to a list of Identities
+   */
+  public push: ProcedureMethod<PushBenefitParams, void>;
```
Comment (monitz87) [inline src/api/entities/DividendDistribution/index.ts:115]: I think `pay` is a better name for this

File: src/api/entities/DividendDistribution/index.ts [hunk 2]
```diff
@@ -95,13 +97,23 @@ export class DividendDistribution extends CorporateAction {
...
+    );
   }
 
   /**
    * Claim the dividends corresponding to the current Identity
    */
   public claim: ProcedureMethod<void, void>;
 
+  /**
+   * Push the capital distribution benefit to a list of Identities
```
Comment (monitz87) [inline src/api/entities/DividendDistribution/index.ts:113]: Transfer the corresponding share of the dividends to a list of Identities

File: src/api/procedures/pushBenefit.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import { DividendDistribution, Identity, PolymeshError, Procedure } from '~/internal';
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:1]: Please call this `payDividends` (rename the file, procedure and test file)

File: src/api/procedures/__tests__/pushBenefit.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+import { getAuthorization, Params, preparePushBenefit } from '~/api/procedures/pushBenefit';
+import { Context, DividendDistribution } from '~/internal';
+import { dsMockUtils, entityMockUtils, procedureMockUtils } from '~/testUtils/mocks';
+import { Mocked } from '~/testUtils/types';
+import { RoleType, TxTags } from '~/types';
+import { PolymeshTx } from '~/types/internal';
+import * as utilsConversionModule from '~/utils/conversion';
+
+describe('pushBenefit procedure', () => {
```
Comment (monitz87) [inline src/api/procedures/__tests__/pushBenefit.ts:12]:
```suggestion
describe('payDividends procedure', () => {
```

File: src/api/procedures/pushBenefit.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  if (expiryDate && expiryDate < now) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'The Distribution has already expired',
+      data: {
+        expiryDate,
+      },
+    });
+  }
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:57]: These same validations are being made in the `claimDividends` procedure. They should probably be refactored into an `assertDistributionOpen` util

File: src/api/procedures/pushBenefit.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  const rawDids = targets.map(target => {
+    if (typeof target === 'string') {
+      return stringToIdentityId(target, context);
+    }
+    return stringToIdentityId(target.did, context);
+  });
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:64]: `const rawDids = targets.map(target =>  stringToIdentityId(signerToString(target), context));`

File: src/api/procedures/pushBenefit.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  holderPaidList.forEach((holderPaid, i) => {
+    if (boolToBoolean(holderPaid)) {
+      alreadyClaimedList.push(i);
+    }
+  });
+
+  if (alreadyClaimedList.length) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'Some of the supplied targets has already claimed their benefits',
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:82]:
```suggestion
      message: 'Some of the supplied Identities have already either been paid or claimed their share of the Distribution',
```

File: src/api/procedures/pushBenefit.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+/**
+ * @hidden
+ */
+export type Params = { distribution: DividendDistribution } & PushBenefitParams;
+
+/**
+ * @hidden
+ */
+export async function preparePushBenefit(
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:24]: Should also throw an error if some of the identities aren't targets of the distribution

File: src/api/procedures/pushBenefit.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  const rawDids = targets.map(target => {
+    if (typeof target === 'string') {
+      return stringToIdentityId(target, context);
+    }
+    return stringToIdentityId(target.did, context);
+  });
+
+  const rawCaId = corporateActionIdentifierToCaId({ ticker, localId }, context);
+
+  const holderPaidList = await Promise.all(
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:68]: Use multi for this

File: src/api/procedures/pushBenefit.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    }
+  });
+
+  if (alreadyClaimedList.length) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'Some of the supplied targets has already claimed their benefits',
+      data: {
+        targets: alreadyClaimedList.map(
+          i => new Identity({ did: identityIdToString(rawDids[i]) }, context)
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:85]:
```suggestion
          i => new Identity({ did: signerToString(targets[i]) }, context)
```

File: src/api/procedures/pushBenefit.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    return stringToIdentityId(target.did, context);
+  });
+
+  const rawCaId = corporateActionIdentifierToCaId({ ticker, localId }, context);
+
+  const holderPaidList = await Promise.all(
+    rawDids.map(rawDid => query.capitalDistribution.holderPaid([rawCaId, rawDid]))
+  );
+
+  const alreadyClaimedList: number[] = [];
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:72]: You could store the identities in this array instead of the indexes

File: src/api/procedures/pushBenefit.ts [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+import { ProcedureAuthorization } from '~/types/internal';
+import { tuple } from '~/types/utils';
+import {
+  boolToBoolean,
+  corporateActionIdentifierToCaId,
+  identityIdToString,
+  stringToIdentityId,
+} from '~/utils/conversion';
+
+export interface PushBenefitParams {
```
Comment (monitz87) [inline src/api/procedures/pushBenefit.ts:12]: PayDividendsParams

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/__tests__/payDividends.ts [hunk 13]
```diff
@@ -145,31 +177,43 @@ describe('pushBenefit procedure', () => {
...
+          entityMockUtils.getIdentityInstance({ did: identityDid })
+        ),
+        treatment: TargetTreatment.Include,
+      },
       expiryDate,
       paymentDate,
     });
-    holderPaidStub.resolves(dsMockUtils.createMockBool(true));
+
+    sinon.stub(utilsConversionModule, 'boolToBoolean').returns(true);
```
Comment (monitz87) [inline src/api/procedures/__tests__/payDividends.ts:190]: This shouldn't be necessary

File: src/api/procedures/payDividends.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+      expiryDate,
+    },
+    targets,
+  } = args;
+
+  assertDistributionOpen(paymentDate, expiryDate);
+
+  const excluded: Identity[] = [];
+  targets.forEach(target => {
+    const targetdDid = signerToString(target);
```
Comment (monitz87) [inline src/api/procedures/payDividends.ts:52]: typo

File: src/utils/internal.ts [hunk 15]
```diff
@@ -477,3 +477,35 @@ export function assertFormatValid(address: string, ss58Format: number): void {
     });
   }
 }
+
+/**
+ * @hidden
+ */
+export function assertDistributionOpen(paymentDate: Date, expiryDate: Date | null): void {
```
Comment (monitz87) [inline src/utils/internal.ts:484]: This should be in procedure utils

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/__tests__/payDividends.ts [hunk 16]
```diff
@@ -170,6 +170,8 @@ describe('payDividends procedure', () => {
     const dids = ['someDid', 'otherDid'];
     const targets = [dids[0], entityMockUtils.getIdentityInstance({ did: dids[1] })];
 
+    sinon.stub(utilsConversionModule, 'boolToBoolean').returns(true);
```
Comment (monitz87) [inline src/api/procedures/__tests__/payDividends.ts:173]: Why is this necessary?

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/__tests__/payDividends.ts (hunk 16 above)
Comment (shuffledex) [reply src/api/procedures/__tests__/payDividends.ts:173]: `should throw an error if some of the supplied targets has already claimed their benefits` brokes if I remove that stub.

Comment (monitz87) [reply src/api/procedures/__tests__/payDividends.ts:173]: Then you're not returning the proper values in the `holderPaid` stub

**Review by monitz87 — APPROVED** [review]
